test/storage/bwtree_index_test.cpp

```diff
@@ -0,0 +1,1204 @@
+    data.reserve(key_schema.size());
+    for (const auto &key : key_schema) {
+      auto key_type = key.GetType();
+      const auto type_size = type::TypeUtil::GetTypeSize(key_type);
+      int64_t rand_int;
+
+      switch (key_type) {
+        case type::TypeId::TINYINT:
+          rand_int = static_cast<int64_t>(static_cast<int8_t>(rng(*generator)));
+          break;
+        case type::TypeId::SMALLINT:
+          rand_int = static_cast<int64_t>(static_cast<int16_t>(rng(*generator)));
+          break;
+        case type::TypeId::INTEGER:
+          rand_int = static_cast<int64_t>(static_cast<int32_t>(rng(*generator)));
+          break;
+        case type::TypeId::BIGINT:
+          rand_int = static_cast<int64_t>(static_cast<int64_t>(rng(*generator)));
+          break;
+        default:
+          throw std::runtime_error("Invalid compact ints key schema.");
+      }
+
+      auto pr_offset = static_cast<uint16_t>(oid_offset_map.at(key.GetOid()));
+      auto attr = pr->AccessForceNotNull(pr_offset);
+      std::memcpy(attr, &rand_int, type_size);
+      data.emplace_back(rand_int);
+    }
+    return data;
+  }
+
+  /**
+   * Modifies a random column of the projected row with the given probability.
+   * Returns true and updates reference if modified, false otherwise.
+   */
+  template <typename Random>
+  bool ModifyRandomColumn(const IndexMetadata &metadata, storage::ProjectedRow *pr, byte *reference, float probability,
+                          Random *generator) {
+    std::bernoulli_distribution coin(probability);
+
+    if (coin(*generator)) {
+      const auto &oid_offset_map = metadata.GetKeyOidToOffsetMap();
+      const auto &key_schema = metadata.GetKeySchema();
+      std::uniform_int_distribution<uint16_t> rng(0, static_cast<uint16_t>(key_schema.size() - 1));
+      const auto column = rng(*generator);
+
+      uint16_t offset = 0;
+      for (uint32_t i = 0; i < column; i++) {
+        offset = static_cast<uint16_t>(offset + type::TypeUtil::GetTypeSize(key_schema[i].GetType()));
+      }
+
+      const auto type = key_schema[column].GetType();
+      const auto type_size = type::TypeUtil::GetTypeSize(type);
+
+      auto pr_offset = static_cast<uint16_t>(oid_offset_map.at(key_schema[column].GetOid()));
+      auto attr = pr->AccessForceNotNull(pr_offset);
+      auto *old_value = new byte[type_size];
+      std::memcpy(old_value, attr, type_size);
+      // force the value to change
+      while (std::memcmp(old_value, attr, type_size) == 0) {
+        WriteRandomAttribute(key_schema[column], attr, reference + offset, generator);
+      }
+      delete[] old_value;
+      return true;
+    }
+
+    return false;
+  }
+
+  /**
+   * Tests:
+   * 1. Scan -> Insert -> Scan -> Delete -> Scan
+   * 2. If primary key or unique index, 2 x Conditional Insert -> Scan -> Delete -> Scan
+   */
+  void BasicOps(Index *const index) {
+    // instantiate projected row and key
+    const auto &metadata = index->metadata_;
```

Comment:
This requires this test class to be a friend class of Index. Maybe we can pass in metadata object since the test controls the creation of metadata. This way, we can separate the friend class, which is only for testing, from the Index class.

Reply:
Inside the builder tests, though, we are not constructing the metadata object directly. The metadata concept is supposed to be purely internal and hidden from the world, and therefore there are no accessors either. We would need to add a friend elsewhere, do you think this is worth pursuing?

Reply:
If there is no way around it, then I am ok with the current implementation.